### DISCUSSION TITLE
ASP.NET server + `HttpClient` over custom `Stream`(s)

### DISCUSSION THREAD

**Discussion by lostmsu** [dotnet/aspnetcore#53572] [Q&A]
How do I configure my ASP.NET app to use a custom transport mechanism?

E.g. on the server side I have an event that fires whenever some client is connected and passes a `Stream` object. I want my ASP.NET app to respond to HTTP requests sent over that `Stream` instead of the usual TCP.

And on the client side I need to configure `HttpClient` so that it called my transport mechanism's `Stream Connect(string host)` method.

The transport mechanism abstractions involved may vary, but ideally I want minimal modifications to the ASP.NET app and keep `HttpClient` as the access abstraction. Ideally the client should be able to send multiple interleaved requests over the same connection.

**Comment by lostmsu** [comment]
OK, I figured I can implement `IConnectionListenerFactory` similar to https://github.com/dotnet/aspnetcore/blob/v8.0.1/src/Servers/Kestrel/Transport.NamedPipes/src/Internal/NamedPipeTransportFactory.cs

What remains unclear is the `HttpClient` side.

**Comment by davidfowl** [comment]
It's your lucky day, I happen to have implemented this:

Server:
https://github.com/davidfowl/YarpTunnelDemo/tree/main/Backend/Transport

Client:
https://github.com/davidfowl/YarpTunnelDemo/blob/main/Frontend/Tunnel/TunnelClientFactory.cs

There's some YARP goop in there but for HttpClient you can do:

```
var client = new HttpClient(new SocketsHttpHandler { ConnectCallback = .. } )
```

That connect callback allows you to return a custom stream. See https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.connectcallback?view=net-7.0
